File: bumperbot_vision/bumperbot_vision/feature_detector.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import Image, CameraInfo
from cv_bridge import CvBridge
import numpy as np

import cv2
import message_filters

class FeatureDetector(Node):
    def __init__(self):
        super().__init__('feature_detector')

        self.prev_image = None
        self.prev_keypoints = None
        self.prev_descriptors = None

        self.bridge = CvBridge()

        # Feature detection and matching
        self.detector = cv2.ORB_create()
        self.matcher = cv2.BFMatcher(cv2.NORM_HAMMING)

        # Subscribers
        image_sub = message_filters.Subscriber(
            self, Image, "left_camera/image_raw", qos_profile=qos_profile_sensor_data)
        info_sub = message_filters.Subscriber(
            self, CameraInfo, "left_camera/camera_info", qos_profile=qos_profile_sensor_data)

        self.approximate_sync = message_filters.ApproximateTimeSynchronizer(
            [image_sub, info_sub], queue_size=10, slop=0.2)
        self.approximate_sync.registerCallback(self.image_callback)

        self.matches_pub = self.create_publisher(
            Image, "matches", qos_profile_sensor_data)

    def image_callback(self, image_msg, info_msg):
        # Camera intrinsic parameters
        self.K = np.array([
            [info_msg.k[0], info_msg.k[1], info_msg.k[2]],
            [info_msg.k[3], info_msg.k[4], info_msg.k[5]],
            [info_msg.k[6], info_msg.k[7], info_msg.k[8]]
        ], dtype=np.float64)

        # Convert ROS Image to OpenCV
        try:
            current_image = self.bridge.imgmsg_to_cv2(image_msg, "bgr8")
        except Exception as e:
            self.get_logger().error(f"Failed to convert image: {e}")
            return

        # Feature detection
        keypoints = self.detector.detect(current_image, None)
        
        # Compute descriptors
        keypoints, descriptors = self.detector.compute(current_image, keypoints)

        if self.prev_image is None:
            self.prev_image = current_image
            self.prev_keypoints = keypoints
            self.prev_descriptors = descriptors
            return

        if descriptors is None or len(descriptors) == 0:
             return

        # Feature matching
        matches = self.matcher.match(self.prev_descriptors, descriptors)

        # Sort matches by distance
        matches = sorted(matches, key=lambda x: x.distance)

        if not matches:
            return

        min_dist = matches[0].distance
        
        # Filter good matches
        good_matches = []
        for match in matches:
            if match.distance <= max(2 * min_dist, 30.0):
                good_matches.append(match)

        # Draw matches
        img_match = cv2.drawMatches(
            self.prev_image, self.prev_keypoints,
            current_image, keypoints,
            good_matches, None
        )

        # Publish matches image
        try:
            msg_matches = self.bridge.cv2_to_imgmsg(img_match, "bgr8")
            msg_matches.header.frame_id = image_msg.header.frame_id
            self.matches_pub.publish(msg_matches)
        except Exception as e:
             self.get_logger().error(f"Failed to publish matches: {e}")

        R, t = self.estimatePose(keypoints, good_matches)
        self.get_logger().info(f"Estimated Rotation:\n{R}\nEstimated Translation:\n{t}")

        # Update previous frame data
        self.prev_image = current_image
        self.prev_keypoints = keypoints
        self.prev_descriptors = descriptors

    def estimatePose(self, curr_keypoints, matches):
        points_1 = np.array([self.prev_keypoints[m.queryIdx].pt for m in matches], dtype=np.float32)
        points_2 = np.array([curr_keypoints[m.trainIdx].pt for m in matches], dtype=np.float32)
        
        # Calculate the Essential Matrix
        essential_matrix, _ = cv2.findEssentialMat(points_1, points_2, self.K)

        # Calculate the Homography Matrix
        homography_matrix, _ = cv2.findHomography(points_1, points_2, cv2.RANSAC, 3)
        _, R, t, _ = cv2.recoverPose(essential_matrix, points_1, points_2, self.K)
        return R, t

def main(args=None):
    rclpy.init(args=args)
    feature_detector = FeatureDetector()
    rclpy.spin(feature_detector)
    feature_detector.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: bumperbot_vision/src/feature_detector.cpp
#include "bumperbot_vision/feature_detector.hpp"

#include "cv_bridge/cv_bridge.hpp"

using namespace std::placeholders;

namespace bumperbot_vision
{
FeatureDetector::FeatureDetector(const std::string & name)
: Node(name)
{
  approximate_sync_ = std::make_shared<ApproximateSync>(
    ApproximatePolicy(10), image_sub_, info_sub_);
  approximate_sync_->registerCallback(
    std::bind(&FeatureDetector::imageCallback, this, _1, _2));
  image_transport::TransportHints hints{this};
  const auto sensor_data_qos = rclcpp::SensorDataQoS().get_rmw_qos_profile();

  image_sub_.subscribe(
    this, "left_camera/image_raw", hints.getTransport(), sensor_data_qos);
  info_sub_.subscribe(this, "left_camera/camera_info", sensor_data_qos);

  matches_pub_ =
    image_transport::create_publisher(this, "matches", rmw_qos_profile_sensor_data);

  detector_ = cv::ORB::create();
  descriptor_ = cv::ORB::create();
  matcher_ = cv::DescriptorMatcher::create("BruteForce-Hamming");
}

void FeatureDetector::imageCallback(
  const sensor_msgs::msg::Image::ConstSharedPtr & image_msg,
  const sensor_msgs::msg::CameraInfo::ConstSharedPtr & info_msg)
{
  // Camera Intrinsics
  K_ = (cv::Mat_<double>(3, 3) << 
    info_msg->k[0], info_msg->k[1], info_msg->k[2],
    info_msg->k[3], info_msg->k[4], info_msg->k[5],
    info_msg->k[6], info_msg->k[7], info_msg->k[8]);

  cv::Mat img = cv_bridge::toCvCopy(image_msg, image_msg->encoding)->image;
  std::vector<cv::KeyPoint> curr_keypoints;
  detector_->detect(img, curr_keypoints);
  cv::Mat curr_descriptor;
  descriptor_->compute(img, curr_keypoints, curr_descriptor);

  if (prev_image_.empty()) {
    // Initialize if first frame
    prev_image_ = img.clone();
    prev_keypoints_ = curr_keypoints;
    prev_descriptor_ = curr_descriptor.clone();
    return;
  }

  std::vector<cv::DMatch> matches, good_matches;
  matcher_->match(prev_descriptor_, curr_descriptor, matches);

  auto min_max = std::minmax_element(matches.begin(), matches.end(),
    [](const cv::DMatch & m1, const cv::DMatch & m2) { return m1.distance < m2.distance; });
  double min_dist = min_max.first->distance;

  for(const auto& m : matches) {
    if(m.distance <= std::max(2 * min_dist, 30.0)) {
      good_matches.push_back(m);
    }
  }

  cv::Mat img_match;
  cv::drawMatches(prev_image_, prev_keypoints_, img, curr_keypoints, good_matches, img_match);

  // Publish the matches image
  auto msg_matches = cv_bridge::CvImage(std_msgs::msg::Header(), "rgb8", img_match).toImageMsg();
  msg_matches->header.frame_id = image_msg->header.frame_id;
  matches_pub_.publish(*msg_matches);

  cv::Mat R, t;
  estimatePose(curr_keypoints, good_matches, R, t);
  RCLCPP_INFO_STREAM(get_logger(), "Estimated Rotation:\n" << R << "\nEstimated Translation:\n" << t);

  prev_image_ = img.clone();
  prev_keypoints_ = curr_keypoints;
  prev_descriptor_ = curr_descriptor.clone();
}

void FeatureDetector::estimatePose(
  const std::vector<cv::KeyPoint> & curr_keypoints,
  const std::vector<cv::DMatch> & matches,
  cv::Mat & R, cv::Mat & t)
{
  std::vector<cv::Point2f> points_1;
  std::vector<cv::Point2f> points_2;

  for (size_t i = 0; i < matches.size(); i++){
    points_1.push_back(prev_keypoints_.at(matches.at(i).queryIdx).pt);
    points_2.push_back(curr_keypoints.at(matches.at(i).trainIdx).pt);
  }

  // Calculate the Essential Matrix
  cv::Mat essential_matrix = cv::findEssentialMat(points_1, points_2, K_);

  // Calculate the Homography Matrix
  cv::Mat homography_matrix;
  homography_matrix = cv::findHomography(points_1, points_2, cv::RANSAC, 3);
  cv::recoverPose(essential_matrix, points_1, points_2, K_, R, t);
}
}  // namespace bumperbot_vision

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<bumperbot_vision::FeatureDetector>("feature_detector"));
  rclcpp::shutdown();
  return 0;
}